python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,362 @@
+    pass
+
+
+class LayerNode:
+    """Defines a node in the `LayerGraph`"""
+
+    def __init__(
+            self,
+            name: str,
+            layer: layers.Layer = None,
+            dependent_children: List[LayerNode] = None,
+            inputs: List[str] = None,
+            inputs_as_list: bool = False
+    ):
+        """
+        Constructor to create a node
+
+        Parameters
+        ----------
+        name: str
+            Name of the node
+        layer: instance of type tf.keras.layers.Layer
+            Node operation, i.e. Layer for forward prop
+        dependent_children: list of LayerNode
+            All downstream nodes which depend on this node's output
+        inputs: list of str
+            All inputs to the `layer` to be passed during forward prop
+        inputs_as_list: bool
+            Boolean to indicate if the set of input tensors to the layer should be converted to list
+        """
+        self.name = name
+        self.layer = layer
+        self.dependent_children = dependent_children if dependent_children is not None else []
+        self.inputs = inputs
+        self.is_input_node = True if inputs is None else False
+        self.inputs_as_list = inputs_as_list
+
+    def __str__(self):
+        """Node is identified by its name"""
+        return self.name
+
+    def __repr__(self):
+        """Get the obj representation"""
+        return self.name
+
+
+class LayerGraph:
+    """Defines a DAG of `LayerNode`"""
+
+    def __init__(
+            self,
+            layer_ops: Dict[str, Union[str, bool, layers.Layer]],
+            inputs: List[str]
+    ):
+        """
+        Constructor to create a Graph. It creates a dependency graph and identifies the output node.
+
+        Parameters
+        ----------
+        layer_ops: dict
+            Dictionary which contains information for all the layer nodes in the graph. Expected schema:
+            {
+                "NodeName":  # Name of the layer node
+                {
+                    "inputs": ["InputNodeName1", "InputNodeName2"],  # Name of the input layer nodes for this layer
+                    "op": keras.layers.Layer.ChildClass(),  # Layer instance which defines the operation of the node
+                    "aslist": True  # If inputs to the "op" should be a list of tensors
+                }
+            }
+        inputs: list of str
+            All inputs to the model from the interaction model
+        """
+        self.nodes = self.create_nodes(layer_ops, inputs)
+        self.create_dependency_graph()
+        # TODO: Add graph visualization for easier debugging
+        output_nodes = self.get_output_nodes()
+        if len(output_nodes) == 0:
+            raise CycleFoundException("No output nodes found because of cycle in DAG")
+        elif len(output_nodes) > 1:
+            raise NotImplementedError(f"1 output node expected, found {len(output_nodes)}: {output_nodes}")
```

Comment:
@arvindsrikantan You could also create a `DanglingNodeFound` exception while you are at it.

Reply:
Dangling nodes will be considered input nodes. Also, we cannot create a non input node without inputs. The idea here is that we might want to eventually support multiple outputs, so I feel NotImplementedError is more appropriate.

Reply:
> Dangling nodes will be considered input nodes

I'm not sure I understand this.

Reply:
All nodes which are dangling will have two properties:
1. They have no inputs (Feature layer nodes) -> In this case, these nodes have is_input_node=True. These are not dangling nodes
2. They have no dependent children (output nodes) -> In this case, we throw NotImplementedError because we want to eventually support multiple outputs for the model

If in case we have a non-feature layer node with 0 inputs and 0 dependent children, it will be ignored similar to a feature node which is not used by any layer which should be okay I feel